libafl/src/events/mod.rs

```diff
@@ -586,80 +596,129 @@ pub trait HasEventManagerId {
-pub struct NopEventManager<S> {
-    phantom: PhantomData<S>,
+/// A tuple of custom buffer handlers
+pub trait CustomBufHandlerTuple<S> {
+    /// Attempt to handle the custom buffer, breaking at the first successful usage
+    fn handle_all(&mut self, state: &mut S, name: &str, buf: &[u8]) -> Result<(), Error>;
 }
 
-impl<S> NopEventManager<S> {
-    /// Creates a new [`NopEventManager`]
-    #[must_use]
-    pub fn new() -> Self {
-        NopEventManager {
-            phantom: PhantomData,
+impl<S> CustomBufHandlerTuple<S> for () {
+    fn handle_all(&mut self, _state: &mut S, _name: &str, _buf: &[u8]) -> Result<(), Error> {
+        Ok(())
+    }
+}
+
+impl<Head, Tail, S> CustomBufHandlerTuple<S> for (Head, Tail)
+where
+    Head: CustomBufHandler<S>,
+    Tail: CustomBufHandlerTuple<S>,
+{
+    fn handle_all(&mut self, state: &mut S, name: &str, buf: &[u8]) -> Result<(), Error> {
+        if let CustomBufEventResult::Next = self.0.handle(state, name, buf)? {
+            self.1.handle_all(state, name, buf)
+        } else {
+            Ok(())
         }
     }
 }
 
-impl<S> Default for NopEventManager<S> {
-    fn default() -> Self {
-        Self::new()
+#[allow(deprecated)]
+impl<S> CustomBufHandlerTuple<S> for Vec<Box<CustomBufHandlerFn<S>>> {
+    fn handle_all(&mut self, state: &mut S, name: &str, buf: &[u8]) -> Result<(), Error> {
+        for handler in self {
+            if let CustomBufEventResult::Handled = handler(state, name, buf)? {
+                return Ok(());
+            }
+        }
+        Ok(())
     }
 }
 
-impl<S> UsesState for NopEventManager<S>
-where
-    S: State,
-{
-    type State = S;
+/// An event manager which can handle custom buffers
+pub trait HasCustomBufHandlers<S> {
+    /// The type of the handlers
+    type Handlers: CustomBufHandlerTuple<S>;
+
+    /// Getter for the handlers
+    fn handlers(&self) -> &Self::Handlers;
+    /// Mutable getter for the handlers
+    fn handlers_mut(&mut self) -> &mut Self::Handlers;
+}
+
+/// Utility trait to maintain compatibility with previous versions which strictly used a list of
+/// dynamic custom buf handlers.
+///
+/// This is automatically implemented for the appropriate traits.
+pub trait HasDynamicHandlers<S> {
+    /// Adds a custom buffer handler that will run for each incoming `CustomBuf` event.
+    #[deprecated(
+        note = "Using function pointers may limit your ability to use some stages due to type conflicts regarding the state"
+    )]
+    #[allow(deprecated)]
+    fn add_custom_buf_handler(&mut self, handler: Box<CustomBufHandlerFn<S>>);
 }
 
-impl<S> EventFirer for NopEventManager<S>
+#[allow(deprecated)]
+impl<EM, S> HasDynamicHandlers<S> for EM
 where
-    S: State,
+    EM: HasCustomBufHandlers<S, Handlers = Vec<Box<CustomBufHandlerFn<S>>>>,
 {
-    fn should_send(&self) -> bool {
-        true
+    fn add_custom_buf_handler(&mut self, handler: Box<CustomBufHandlerFn<S>>) {
+        self.handlers_mut().push(handler);
     }
+}
```

Comment:
@domenukk This retains the dyn FnMut behaviour without making it mandatory.

Reply:
why prepend with `Has`?

Reply:
Because the thing which *has* the dynamic handlers is what is added to.

Reply:
I don't like `Has` for traits too much. The trait isn't that it _has_ dynamic handlers but that it calls them or that it allows you to add them or whatever. But I'm not a native speaker, so..

Reply:
Hm, what would you prefer? We can specify that `Has` means that only getters are provided, but then we need to update e.g. `HasTestcase`.

Reply:
HasTestcase means it has a `testcase()` method that hands out a `testcase`. That's all the trait does.

This trait doesn't just have dynamic handlers.
Something like `DynamicBufHandlers` or so?

Reply:
It's the thing that _calls_ the handlers, not the thing that _has_ the handlers, that's my whole point. `Has` is just a lazy name for traits. There are no traits with `Has` prefix in the stdlib for example

Reply:
If you have a better name, I'm open to suggestions. This is pretty consistent with how we've named other things imo :shrug:

Reply:
`SupportsDynamicHandlers`?

Reply:
:+1: Makes sense to me